randovania/games/am2r/generator/pool_creator.py

```diff
@@ -2,11 +2,111 @@
 
 from randovania.game_description.game_description import GameDescription
 from randovania.game_description.game_patches import GamePatches
+from randovania.game_description.resources.pickup_entry import PickupEntry
+from randovania.game_description.resources.pickup_index import PickupIndex
+from randovania.game_description.db.node_identifier import NodeIdentifier
 from randovania.games.am2r.layout.am2r_configuration import AM2RConfiguration
 from randovania.generator.pickup_pool import PoolResults
+from randovania.generator.pickup_pool.pickup_creator import create_am2r_artifact
+from randovania.generator.pickup_pool.pool_creator import _extend_pool_results
 from randovania.layout.base.base_configuration import BaseConfiguration
 
 
 def pool_creator(results: PoolResults, configuration: BaseConfiguration, game: GameDescription,
                  base_patches: GamePatches, rng: Random) -> None:
     assert isinstance(configuration, AM2RConfiguration)
+
+    _extend_pool_results(results, artifact_pool(game, configuration, rng))
+
+
+def artifact_pool(game: GameDescription, configuration: AM2RConfiguration, rng: Random) -> PoolResults:
+    config = configuration.artifacts
+
+    new_assignment: dict[PickupIndex, PickupEntry] = {}
+
+    keys: list[PickupEntry] = [create_am2r_artifact(i, game.resource_database) for i in range(46)]
+
+    keys_to_shuffle = keys[:config.required_artifacts]
+    starting_keys = keys[config.required_artifacts:]
+
+    locations: list[NodeIdentifier] = []
+    if config.prefer_metroids:
+        locations.extend(_METROID_INDICES)
+    if config.prefer_bosses:
+        locations.extend(_BOSS_INDICES)
+
+    item_pool = keys_to_shuffle
+
+    if rng is not None:
+        rng.shuffle(locations)
+        new_assignment = {
+            game.region_list.get_pickup_node(location).pickup_index: key
+            for location, key in zip(locations, keys_to_shuffle)
+        }
+        item_pool = [key for key in keys_to_shuffle if key not in new_assignment.values()]
+
+    return PoolResults(item_pool, new_assignment, starting_keys)
+
+_c = NodeIdentifier.create
+_METROID_INDICES = [
+    # Alphas
+    _c("Main Caves", "Surface Alpha Nest", "Pickup (Alpha)"),
+    _c("Golden Temple", "Exterior Alpha Nest", "Pickup (Alpha)"),
```

Comment:
Disclaimer: I'm not a core dev, which means it's not up to me to decide. I also know that most other games already doing stuff like that and especially this part is basically a copy of dread. Eventually, we better ask someone additional.

I really dislike hard coding stuff like names from the db. This breaks easily as soon as any part is renamed. My suggestion would be to use the extra section and simply filter the nodes according to these optional flags. That would make renaming much easier and less faulty. But as I said, that's more a general problem in randovania IMO and someone else should be asked.

Reply:
I copied dreads implementation for now as a way to get this implemented and playable faster.
I do agree though that the hardcoded nodes should go away for Dread, AM2R and whatever uses them as well. 
Due to tests, if something would change in the DB, it would complain here so a "something broken and generation is impossible now" shouldn't happen, but at minimum for when the "place DNA at normal pickups" option gets implemented this basically *has* to get refactored because I wouldn't want have all 88 pickup nodes in here.

Reply:
Great refactor idea, did it for Dread https://github.com/randovania/randovania/pull/4538

Reply:
Addressed now as well.